src/assets/scripts/client/trafficGenerator/buildPreSpawnAircraft.js

```diff
@@ -1,83 +1,271 @@
+        const nextWaypointModel = waypointModelList[i];
+        const distanceToNextWaypoint = previousWaypointModel.calculateDistanceToWaypoint(nextWaypointModel);
+        totalDistanceTraveled += distanceToNextWaypoint;
+
+        waypointOffsetMap.push(totalDistanceTraveled);
+    }
+
+    return waypointOffsetMap;
+}
+
+/**
+ * Create an array of the various altitude restrictions along the route, and the offset distances at which they exist
+ *
+ * Note that unlike the `waypointOffsetMap`, the altitude offset array DOES NOT mirror the indices
+ * of the `waypointModelList`. Only offsets and altitudes of altitude-restricted waypoints are included.
+ *
+ * Also note that this is designed to retrieve bottom altitudes used in descent. It could probably
+ * be reused for departures, just by doing a Math.max() instead of Math.min() below
+ *
+ * @function _calculateAltitudeOffsets
+ * @param waypointModelList {array<WaypointModel>} array of all waypoints along the route
+ * @param waypointOffsetMap {array<number>} offset distance from spawn point to each waypoint
+ * @return {array<array<number>>} [[offsetDistance, altitude], [offsetDistance, altitude], ...]
+ */
+export function _calculateAltitudeOffsets(waypointModelList, waypointOffsetMap) {
+    const altitudeOffsets = [];
+
+    for (let i = 0; i < waypointModelList.length; i++) {
+        const waypointModel = waypointModelList[i];
+
+        if (!waypointModel.hasAltitudeRestriction) {
+            continue;
+        }
+
+        const altitudes = _without([waypointModel.altitudeMaximum, waypointModel.altitudeMinimum], -1);
+
+        altitudeOffsets.push([
+            waypointOffsetMap[i],
+            Math.min(...altitudes)
+        ]);
+    }
+
+    return altitudeOffsets;
+}
+
+/**
+ * Calculate the interpolated altitude along the glidepath at the given distance along the route from the spawn point
+ *
+ * @function _calculateAltitudeAtOffset
+ * @param altitudeOffsets {array<array<number>>} information about location of altitude restrictions
+ * @param offsetDistance {number} distance along route at which we want to know the ideal spawn altitude
+ * @return {number} ideal spawn altitude, in feet (rounded up to nearest thousand)
+ */
+export function _calculateAltitudeAtOffset(altitudeOffsets, offsetDistance) {
+    const indexOfDistance = 0;
+    const indexOfAltitude = 1;
+    const indexOfNextAltitudeRestriction = _findIndex(altitudeOffsets, (altitudeOffset) => {
+        return altitudeOffset[indexOfDistance] >= offsetDistance;
+    });
+    let indexOfPreviousAltitudeRestriction = indexOfNextAltitudeRestriction - 1;
+
+    if (indexOfPreviousAltitudeRestriction < -1) {
+        indexOfPreviousAltitudeRestriction = altitudeOffsets.length - 1;
+    }
+
+    if (indexOfNextAltitudeRestriction < 0) { // no restrictions ahead
+        if (indexOfPreviousAltitudeRestriction < 0) { // no restrictions ahead or behind
+            return undefined;
```

Comment:
why `undefined` here instead of `null`?

Reply:
Looked into it, this is actually overly defensive. What happens is when there are no restrictions, `_calculateIdealSpawnAltitudeAtOffset()` _creates_ a restriction to cross the airspace boundary at the highest owned altitude. It then calls `_calculateAltitudeAtOffset()` and provides an array of the restrictions' offsets and altitudes. Thus, the restrictions array should never be empty if called from this function, but `_calculateAltitudeAtOffset()` is checking it anyway. In the event that it _does_ receive an empty restriction array, it returns undefined as the altitude. Since we're pre-checking it though, it should actually probably throw instead of returning a value which could mess stuff up.

Reply:
Tests updated to reflect this.